### ISSUE TITLE
Middleware cannot abort asynchronous action

### ISSUE THREAD

**Issue by d4rky-pl** [mobxjs/mobx-state-tree#621]
Due to the way asynchronous actions are being called, there's no way to skip all asynchronous actions without also skipping synchronous ones.

As [the documentation states](https://github.com/mobxjs/mobx-state-tree/blob/master/docs/middleware.md)

> If next(call) is not invoked by your middleware, the action will be aborted and not actually executed.

and

> Note: for middleware, it is extremely important that next(action) is called for asynchronous actions (call.type !== "action"), otherwise the generator will remain in an unfinished state forever

Unfortunately during the `action` step we have no way of detecting if the action is asynchronous and on `flow_spawn` we cannot abort by skipping the `next(action)` because it makes the action hang indefinitely.

**Comment by robinfehr** [comment]
brainstorming / having a look at the code properly again needed:
- maybe the call/ action needs to contain the information
- maybe decorators could help
- maybe even something different than flow

**Comment by mweststrate** [comment]
Could we introduce a global flag, so that we know in middleware whether a
flow spawned? Eg

```
function middleware(call, next) {
    const res = next(call)
    if (mst.flowSpawned === true) {
      res.then(something)
   }
   return res
}
```

Op wo 21 mrt. 2018 om 14:09 schreef Rob <notifications@github.com>:

**Comment by robinfehr** [comment]
@mweststrate such global flag would only be a closure during the runtime of the flow right?
if yes that would be aligned with the decoratorFactory idea we had.

**Comment by mweststrate** [comment]
Yes. Undefined immediately after running middleware handlers.

Op wo 21 mrt. 2018 17:32 schreef Rob <notifications@github.com>:

**Comment by d4rky-pl** [comment]
I'd be ok with either A) an `async` flag on action or B) extra event (`flow_start`) that happens before generator is spawned. Both would fix the issue perfectly :)

**Comment by jayarjo** [comment]
Was this ever resolved in any way?

**Comment by jamonholmgren** [comment]
Fixed in v5.1.8: https://github.com/mobxjs/mobx-state-tree/releases/tag/v5.1.8
